### ISSUE TITLE
Allow Alice to publish many BlockchainArrangements with a single transaction

### ISSUE THREAD

**Issue by jMyles** [nucypher/nucypher#563]
*tl;dr: Presently, Alice conducts `n` transactions (and pays for gas for each, independently) for any `Policy`.*

### The basic mechanics
When a dapp calls `Alice.grant(...)`, several things happen behind the scenes so that our dapp audience need not consider how much logic is needed to actually make a `Policy`.  Most of these are not relevant for this issue (sampling, finding `Ursulas`, making `Arrangements`, publishing the `TreasureMap`, etc).  The important part here is this:

`BlockchainPolicy.publish(...)` runs a `while` loop, calling `publish(...)` on each arrangement (which is a `BlockchainArrangement` until all are in the `_enacted_arrangements` bucket.

This causes `<Arrangement()>.policy_agent.contract.functions.createPolicy(...)` to be called inside this loop.

Here's the signature for `createPolicy`:
```
    function createPolicy(
        bytes16 _policyId,
        uint16 _numberOfPeriods,
        uint256 _firstPartialReward,
        address[] _nodes
    )
```

As you can see, `createPolicy` takes an array of address (and is thus a number match for `BlockchainPolicy`, even though we're calling it here from `BlockchainArrangement`).

### But wasn't this for anonymity?

The short answer is 'yes'.  But the long answer is more complex and more meaningful.

It has always been a stated goal to provide for anonymity in our logic, so that when we (and the larger blockchain ecosystem) is ready for anonymous deployment, we don't have to rework all of our assumptions and metaphors to make it work.  Great!

However, it has also been a stated goal to launch this project to mainnet before anonymity is available. 

**It is already possible for `Policy-related` Ursulas to find one another by dint of publicly available data produced by our network, including on the blockchain.** 

As such, it probably makes sense to optionally allow Alice to place many `Arrangements` in a single transaction, as is already afforded in the contract.  In fact, it's possible that Alice can place the Arrangements from more than one `Policy` in a single transaction!

### The bottom-line
NuCypher needs to be affordable to gain traction - we all agree about this.  One of the easiest ways that we can bring costs down for a large number of dapp use cases is to reduce the number of transactions Alice needs to conduct from `n` per `Policy` to `1` per `Policy` or even further.

**Comment by KPrasch** [comment]
@szotov - We can call this closed?

**Comment by szotov** [comment]
I think, yes we can

**Comment by cygnusv** [comment]
@szotov Just wondering: Is this closed because Alice can now publish many arrangements with a single transaction?

**Comment by szotov** [comment]
@cygnusv - In blockchain it was a long time ago, and as far I know it was added into `PolicyAgent` too

**Comment by jMyles** [comment]
Wait now, let's not just "call it closed" - let's link the commit that closed it.

**Comment by KPrasch** [comment]
Phew! That will be a dig!

**Comment by cygnusv** [comment]
Yes, please, because that went under my radar.

**Comment by jMyles** [comment]
In any case, it's not fixed on master as of this moment:

https://github.com/nucypher/nucypher/blob/3ed04cef945987eaa39959000b1fe2081b7e5265/nucypher/blockchain/eth/policies.py#L246

**Comment by jMyles** [comment]
It appears... although I haven't fully introspected this to ensure that it is indeed a single tx, but it appears to be fixed in #805:

https://github.com/nucypher/nucypher/blob/60e8611babcef0d2f31f4f196a3a6223e9e27cde/nucypher/blockchain/eth/policies.py#L265

**Comment by KPrasch** [comment]
I do not think that is evidence of it not being fixed, in fact all that method does it publish the treasure map.

**Comment by jMyles** [comment]
Here's the commit (26149f8409902ec67b6cd9dc3529142697ba6662):

https://github.com/nucypher/nucypher/commit/26149f8409902ec67b6cd9dc3529142697ba6662

**Comment by KPrasch** [comment]
Hmm, that commit does not really address *that* Issue.

**Comment by jMyles** [comment]
Here's the commit that removed the while loop: https://github.com/nucypher/nucypher/commit/2ebd3fc2d1e87efac8716ef4605ef28efc287074#diff-5b3bd97d75d539a8b1311b82d491cb58

**Comment by KPrasch** [comment]
In order to really find the commit where this is fixed, We'll need to find the solidity changes,

**Comment by KPrasch** [comment]
To be clear: my understanding is that this was fixed a *very* long time ago.

**Comment by jMyles** [comment]
I don't understand what you mean when you say "the solidity changes" - the whole problem was that the solidity already *took* an array, but we were calling it in a loop and passing a single item.  If there were solidity changes required to fix this, it wasn't obvious to me at that time that I opened it (though obviously that doesn't mean it's not so :-) ).

**Comment by jMyles** [comment]
@KPrasch: Did you see my link above?  It's literally not fixed on master at this moment.  It was not fixed a long time ago.  Perhaps you are misreading the issue?

**Comment by KPrasch** [comment]
Wait what? That is not my understanding of the Issue.

**Comment by jMyles** [comment]
I mean... I think it's a fairly clear Issue description - is there part of it to which I can add clarity?  I said:

> As you can see, `createPolicy` takes an array of address (and is thus a number match for `BlockchainPolicy`, even though we're calling it here from `BlockchainArrangement`).

**Comment by KPrasch** [comment]
Well the title of this Issue is "Allow Alice to publish many BlockchainArrangements with a single transaction" - Which is fixed 100%.

**Comment by jMyles** [comment]
...not in master.  In master, this is still the case:

> The important part here is this:

> `BlockchainPolicy.publish(...)` runs a `while` loop, calling `publish(...)` on each arrangement (which is a `BlockchainArrangement` until all are in the `_enacted_arrangements` bucket.

**Comment by jMyles** [comment]
However, it does (at least at first blush) appear fixed in #805 as part of the `BlockchainPolicy` refactor.

**Comment by KPrasch** [comment]
I mean you actually can do it in one transaction, the solidity code allows it - That's what I mean by its fixed.

**Comment by KPrasch** [comment]
The code you point out above does not behave in the way you describe - It does not make multiple transactions.

**Comment by jMyles** [comment]
...that's not what this Issue is.  In the description, I explained that, although the solidity takes an array, `BlockchainPolicy` does not provide a way to publish using that array, and instead calls `createPolicy` repeatedly in a loop.

I really think that if you just skim the description again, this will all be clear.

**Comment by KPrasch** [comment]
That code was "detached" from the rest of the codebase for a while, I'm just trying to point out that the change to make this possible in a single transaction, happened some time ago, but this policy code didn't get updated with it.

**Comment by jMyles** [comment]
Yeah, that's the same thing I linked above.  So indeed, it's not yet fixed on master.  But... looks like it will be when this is merged, yeah?

**Comment by KPrasch** [comment]
Yes, and in this example, `createPolicy` is not being called in a loop, right?

**Comment by jMyles** [comment]
Right.

**Comment by KPrasch** [comment]
Above you said it *was* being called in a loop. but it is infact, not.

**Comment by KPrasch** [comment]
Yes, where is `arrangement.publish` making a TX?

**Comment by jMyles** [comment]
I mean... it *was* until that commit (which both you and I linked).  It seems like there's no question here - the while loop was removed from line 244 in that commit, and that, at least in theory, closes this issue.

**Comment by KPrasch** [comment]
Even *before* this change, `arrangement.publish` did not execute a TX, right?

**Comment by jMyles** [comment]
Yes it did - here: https://github.com/nucypher/nucypher/blob/3ed04cef945987eaa39959000b1fe2081b7e5265/nucypher/blockchain/eth/policies.py#L82

**Comment by KPrasch** [comment]
Ah! There is it. I really could not find it.

**Comment by jMyles** [comment]
Yeah, it was a motherfucker to wrap the head around - that's why I opened this issue with such a thorough description.  ;-)

**Comment by KPrasch** [comment]
Well - Id like to feel understood here: there is a real difference between it being possible in one TX and the policy code implementing the TX correctly.

**Comment by jMyles** [comment]
Yes yes - the issue was always about the latter, and I made clear in the description that the former already appeared possible.

**Comment by jMyles** [comment]
(thought the solidity has changed as well now, but it was actually already possible when this issue was opened)

**Comment by KPrasch** [comment]
Yeah, that's what I thought, and where I was coming from when trying to close this Issue as part of cleanup.

**Comment by jMyles** [comment]
It seems like - although I'm not sure if there's a distinct test for this, and you may want to throw one together - you can mark this fixed in the description of #805.
